### DISCUSSION TITLE
Rethink extras & plugins for docTR

### DISCUSSION THREAD

**Discussion by frgfm** [mindee/doctr#1064] [Ideas]
Hello there :wave:

As docTR grows, numerous features are added. Added code doesn't really impact the end user/production level, however, adding dependencies does (potentially heavier project, but for certain tighter version constraints). So I thought it would be a good occasion to:
- identify all "side features" vs. production/inference features
- assess which dependencies are the heaviest / trickiest
- decide whether we move them as an extra (part of the build command) or plugin

Here are all the side features I think about:
- language detection (#1023) with `langdetect`
- HTML rendering (https://github.com/mindee/doctr/blob/main/doctr/io/html.py) with `weasyprint`
- PDF rendering (https://github.com/mindee/doctr/blob/main/doctr/io/pdf.py) with `pypdfium2`
- Result plotting (https://github.com/mindee/doctr/blob/main/doctr/utils/visualization.py) with `matplotlib` & `mplcursors`
- XML export (https://github.com/mindee/doctr/blob/main/doctr/io/elements.py) using `defusedxml`

What do you think?

This is related to #1024 but not specific to model output features only, so I didn't consider this a duplicate.

**Comment by felixdittrich92** [comment]
Hi @frgfm sounds good to me (i would include the object detection also)

This would be also a good starting point for Onnx to encapsulate the modules as standalone / extras
How do you think should we handle the extra build ? python-doctr[torch-extras]  / python-doctr[torch] ? 
We should think carefully about how we want to handle this, because I think it will change the basic structure of the project :)

**Reply by frgfm** [reply]
which part of object detection? (artefact detection with DL might be considered a side feature but we won't need anything that we don't in core)

Yes the decisions here can have significant consequences :upside_down_face:

**Reply by felixdittrich92** [reply]
I see the object detection as external 'extra' i think the focus is pure OCR :)

**Reply by frgfm** [reply]
by object detection you mean : artefact detection using object detection models? or you mean using cv2 to detect faces, barcodes, etc?

**Reply by felixdittrich92** [reply]
short answer: both  .... everything which is not directly related to OCR
